File: src/main/java/org/apache/prepbuddy/qualityanalyzers/DataType.java
package org.apache.prepbuddy.qualityanalyzers;

import org.apache.prepbuddy.utils.Range;

import java.io.Serializable;
import java.util.*;

/**
 * Data types that can be inferred by inspecting the data in a column
 * <p/>
 * todo :
 * Add the following new types
 * BOOLEAN
 * LATITUDE_LONGITUDE_PAIR
 * CATEGORICAL_INTEGER
 * CATEGORICAL_STRING
 * CREDIT_CARD
 * GENDER
 * HTTP_CODE
 * US_STATE
 * and user defined data types
 */
public enum DataType implements Serializable {
    EMPTY {
        @Override
        public boolean isOfType(List<String> sampleData) {
            List<String> emptyRepresentation = Arrays.asList("N\\A", "\\N", "NULL", "blank", "(blank)", "NaN");
            for (String dataValue : sampleData) {
                if (!emptyRepresentation.contains(dataValue))
                    return false;
            }
            return true;
        }
    },
    INTEGER {
        @Override
        public boolean isOfType(List<String> sampleData) {
            final String EXPRESSION = "^[+-]?\\d+$";
            boolean matches = matchesWith(EXPRESSION, sampleData);
            return matches && extraChecksAreSuccessful(sampleData);
        }

        private boolean extraChecksAreSuccessful(List<String> sampleData) {
            for (String dataPoint : sampleData) {
                if (dataPoint.contains(".")) return false;
            }
            return true;
        }
    },
    URL {
        @Override
        public boolean isOfType(List<String> sampleData) {
            final String EXPRESSION = "^(https?|ftp|file)://[-a-zA-Z0-9+&@#/%?=~_|!:,.;]*[-a-zA-Z0-9+&@#/%=~_|]$";
            return matchesWith(EXPRESSION, sampleData);
        }
    },
    EMAIL {
        @Override
        public boolean isOfType(List<String> sampleData) {
            final String EXPRESSION = "^[_A-Za-z0-9-]+(\\.[_A-Za-z0-9-]+)*@[A-Za-z0-9]+(\\.[A-Za-z0-9]+)*(\\.[A-Za-z]{2,})$";
            return matchesWith(EXPRESSION, sampleData);
        }
    },
    CURRENCY {
        @Override
        public boolean isOfType(List<String> sampleData) {
            final String EXPRESSION = "^(\\p{Sc})(\\d+|\\d+.\\d+)$";
            return matchesWith(EXPRESSION, sampleData);
        }
    },
    DECIMAL {
        @Override
        public boolean isOfType(List<String> sampleData) {
            final String EXPRESSION = "^[+-]?(\\.\\d+|\\d+\\.\\d+)$";
            return matchesWith(EXPRESSION, sampleData);
        }
    },
    SOCIAL_SECURITY_NUMBER {
        @Override
        public boolean isOfType(List<String> sampleData) {
            final String EXPRESSION = "^(\\d{3}-\\d{2}-\\d{4})$";
            return matchesWith(EXPRESSION, sampleData);
        }
    },
    IP_ADDRESS {
        @Override
        public boolean isOfType(List<String> sampleData) {
            final String EXPRESSION = "^(([01]?\\d\\d?|2[0-4]\\d|25[0-5])\\.){3}([01]?\\d\\d?|2[0-4]\\d|25[0-5])$";
            return matchesWith(EXPRESSION, sampleData);
        }
    },
    ZIP_CODE_US {
        @Override
        public boolean isOfType(List<String> sampleData) {
            final String EXPRESSION = "^[0-9]{5}(?:-[0-9]{4})?$";
            return matchesWith(EXPRESSION, sampleData);
        }
    },
    ALPHANUMERIC_STRING {
        @Override
        public boolean isOfType(List<String> sampleData) {
            return false;
        }
    },
    COUNTRY_CODE_2_CHARACTER {
        @Override
        public boolean isOfType(List<String> sampleData) {
            TreeSet<String> char2countryCodes = new TreeSet<>(Arrays.asList(Locale.getISOCountries()));
            return matchInDictionary(sampleData, char2countryCodes);
        }
    },
    COUNTRY_CODE_3_CHARACTER {
        @Override
        public boolean isOfType(List<String> sampleData) {
            Set<String> countryCodes = getISO3Codes();
            return matchInDictionary(sampleData, countryCodes);
        }

        private Set<String> getISO3Codes() {
            String[] isoCountries = Locale.getISOCountries();
            TreeSet<String> countryCodes = new TreeSet<>();
            for (String country : isoCountries) {
                Locale locale = new Locale("", country);
                countryCodes.add(locale.getISO3Country());
            }
            return countryCodes;
        }
    },
    COUNTRY_NAME {
        @Override
        public boolean isOfType(List<String> sampleData) {
            Set<String> countryNames = getCountryNames();
            return matchInDictionary(sampleData, countryNames);
        }

        private Set<String> getCountryNames() {
            String[] isoCountries = Locale.getISOCountries();
            TreeSet<String> countryCodes = new TreeSet<>();
            for (String country : isoCountries) {
                Locale locale = new Locale("", country);
                countryCodes.add(locale.getDisplayCountry());
            }
            return countryCodes;
        }
    }, MOBILE_NUMBER {
        @Override
        public boolean isOfType(List<String> sampleData) {
            final String EXPRESSION = "^(([+]\\d+\\s)|0)?\\d{10}$";
            return matchesWith(EXPRESSION, sampleData);
        }
    }, TIMESTAMP {
        @Override
        public boolean isOfType(List<String> sampleData) {
            final String EXPRESSION = "(\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}\\.\\d{3}x?)";
            return matchesWith(EXPRESSION, sampleData);
        }
    }, LATITUDE {
        @Override
        public boolean isOfType(List<String> sampleData) {
            final String EXPRESSION = "^[-+]?([1-8]?\\d(\\.\\d+)?|90(\\.0+)?)$";
            boolean matched = matchesWith(EXPRESSION, sampleData);
            return matched && extraChecksAreSuccessful(sampleData);
        }

        private boolean extraChecksAreSuccessful(List<String> sampleData) {
            Range range = new Range(-90, 90);
            for (String dataPoint : sampleData) {
                if (!range.contains(Double.parseDouble(dataPoint))) return false;
            }
            return true;
        }
    }, LONGITUDE {
        @Override
        public boolean isOfType(List<String> sampleData) {
            final String EXPRESSION = "^[-+]?(180(\\.0+)?|((1[0-7]\\d)|([1-9]?\\d))(\\.\\d+)?)$";
            boolean matched = matchesWith(EXPRESSION, sampleData);
            return matched && extraChecksAreSuccessful(sampleData);
        }

        private boolean extraChecksAreSuccessful(List<String> sampleData) {
            Range range = new Range(-180, 180);
            for (String dataPoint : sampleData) {
                if (!range.contains(Double.parseDouble(dataPoint))) return false;
            }
            return true;
        }
    }, CATEGORICAL_INTEGER {
        @Override
        public boolean isOfType(List<String> sampleData) {
            int categoricalSize = 2;
            return isCategorical(sampleData,categoricalSize);
        }
    }, CATEGORICAL_STRING {
        @Override
        public boolean isOfType(List<String> sampleData) {
            int categoricalSize = 3;
            return isCategorical(sampleData,categoricalSize);
        }
    };

    protected boolean isCategorical(List<String> sampleData,int categoricalSize) {
        TreeSet<String> tree = new TreeSet<>();
        for (String element : sampleData)
            tree.add(element);
        return tree.size()<=categoricalSize;
    }

    protected boolean matchInDictionary(List<String> sampleData, Set<String> dictionary) {
        double qualifyingLimit = sampleData.size() * 0.75;
        Set<String> lowerCaseDictionary = new TreeSet<>();
        for (String element : dictionary)
            lowerCaseDictionary.add(element.toLowerCase());

        int matchCount = 0;
        for (String element : sampleData)
            if (lowerCaseDictionary.contains(element.toLowerCase()))
                matchCount++;
        return matchCount >= qualifyingLimit;
    }

    public boolean matchesWith(String regex, List<String> samples) {
        int counter = 0;
        int threshold = samples.size() / 2;
        for (String string : samples)
            if (string.matches(regex)) counter++;
        return (counter >= threshold);
    }

    public abstract boolean isOfType(List<String> sampleData);
}

File: src/main/scala/org/apache/datacommons/prepbuddy/qualityanalyzers/DataType.scala
package org.apache.datacommons.prepbuddy.qualityanalyzers

import java.util.Locale


abstract class DataType {
    def isOfType(sampleData: List[String]): Boolean

    protected def matches(regex: String, samples: List[String]): Boolean = {
        val matches: List[String] = samples.filter(_.matches(regex))
        val threshold = Math.round(samples.length * 0.75)
        matches.size > threshold
    }


    protected def matchInDictionary(sampleData: List[String], dictionary: Set[String]): Boolean = {
        val sampleInLowerCase: Set[String] = sampleData.map(_.toLowerCase).toSet
        val qualifyingLimit: Long = Math.round(sampleInLowerCase.size * 0.75)
        val size: Int = dictionary.intersect(sampleInLowerCase).size
        size >= qualifyingLimit
    }

    protected def isCategorical(sampleData: List[String], categoricalSize: Int): Boolean = {
        var set: Set[String] = Set()
        set = set.++:(sampleData)
        set.size <= categoricalSize
    }
}

object ALPHANUMERIC_STRING extends DataType {
    override def isOfType(sampleData: List[String]): Boolean = true
}

object DECIMAL extends DataType {
    override def isOfType(sampleData: List[String]): Boolean = {
        val EXPRESSION: String = "^[+-]?(\\.?\\d+|\\d+\\.\\d+)$"
        matches(EXPRESSION, sampleData)
    }
}

object INTEGER extends DataType {
    override def isOfType(sampleData: List[String]): Boolean = {
        val EXPRESSION: String = "^[+-]?\\d+$"
        matches(EXPRESSION, sampleData)
    }
}

object EMAIL extends DataType {
    override def isOfType(sampleData: List[String]): Boolean = {
        val EXPRESSION: String = "^[_A-Za-z0-9-]+(\\.[_A-Za-z0-9-]+)*@[A-Za-z0-9]+(\\.[A-Za-z0-9]+)*(\\.[A-Za-z]{2,})$"
        matches(EXPRESSION, sampleData)
    }
}

object CURRENCY extends DataType {
    override def isOfType(sampleData: List[String]): Boolean = {
        val EXPRESSION: String = "^(\\p{Sc})(\\d+|\\d+.\\d+)$"
        matches(EXPRESSION, sampleData)
    }
}

object URL extends DataType {
    override def isOfType(sampleData: List[String]): Boolean = {
        val EXPRESSION: String = "^(https?|ftp|file)://[-a-zA-Z0-9+&@#/%?=~_|!:,.;]*[-a-zA-Z0-9+&@#/%=~_|]$"
        matches(EXPRESSION, sampleData)
    }
}

object IP_ADDRESS extends DataType {
    override def isOfType(sampleData: List[String]): Boolean = {
        val EXPRESSION: String = "^(([01]?\\d\\d?|2[0-4]\\d|25[0-5])\\.){3}([01]?\\d\\d?|2[0-4]\\d|25[0-5])$"
        matches(EXPRESSION, sampleData)
    }
}

object ZIP_CODE_US extends DataType {
    override def isOfType(sampleData: List[String]): Boolean = {
        val EXPRESSION: String = "^[0-9]{5}(?:-[0-9]{4})?$"
        matches(EXPRESSION, sampleData)
    }
}

object MOBILE_NUMBER extends DataType {
    override def isOfType(sampleData: List[String]): Boolean = {
        val EXPRESSION: String = "^(([+]\\d+\\s)|0)?\\d{10}$"
        matches(EXPRESSION, sampleData)
    }
}

object SOCIAL_SECURITY_NUMBER extends DataType {
    override def isOfType(sampleData: List[String]): Boolean = {
        val EXPRESSION: String = "^(\\d{3}-\\d{2}-\\d{4})$"
        matches(EXPRESSION, sampleData)
    }
}

object TIMESTAMP extends DataType {
    override def isOfType(sampleData: List[String]): Boolean = {
        val EXPRESSION: String = "(\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}\\.\\d{3}[Z]?)"
        matches(EXPRESSION, sampleData)
    }
}

object LONGITUDE extends DataType {
    override def isOfType(sampleData: List[String]): Boolean = {
        val EXPRESSION: String = "^[-+](180(\\.0+)?|((1[0-7]\\d)|([1-9]?\\d))(\\.\\d+)?)$"
        matches(EXPRESSION, sampleData)
    }
}

object LATITUDE extends DataType {
    override def isOfType(sampleData: List[String]): Boolean = {
        val EXPRESSION: String = "^[-+]([1-8]?\\d(\\.\\d+)?|90(\\.0+)?)$"
        matches(EXPRESSION, sampleData)
    }
}

object COUNTRY_NAME extends DataType {

    override def isOfType(sampleData: List[String]): Boolean = {
        val countryNames: Set[String] = getCountryNames
        matchInDictionary(sampleData, countryNames)
    }

    private def getCountryNames: Set[String] = {
        val isoCountries: Array[String] = Locale.getISOCountries
        val countryList: Array[String] = isoCountries.map((country) => {
            val locale = new Locale("", country)
            locale.getDisplayCountry.toLowerCase
        })
        countryList.toSet
    }
}

object COUNTRY_CODE_3_CHARACTER extends DataType {
    override def isOfType(sampleData: List[String]): Boolean = {
        val countryNames: Set[String] = getCountryCodes
        matchInDictionary(sampleData, countryNames)
    }

    private def getCountryCodes: Set[String] = {
        val isoCountries: Array[String] = Locale.getISOCountries
        val countryList: Array[String] = isoCountries.map((country) => {
            val locale = new Locale("", country)
            locale.getISO3Country.toLowerCase
        })
        countryList.toSet
    }
}

object COUNTRY_CODE_2_CHARACTER extends DataType {
    override def isOfType(sampleData: List[String]): Boolean = {
        val countryNames: Set[String] = Locale.getISOCountries.map(_.toLowerCase).toSet
        matchInDictionary(sampleData, countryNames)
    }
}

object CATEGORICAL_STRING extends DataType {
    override def isOfType(sampleData: List[String]): Boolean = {
        val categoricalSize: Int = 3
        isCategorical(sampleData, categoricalSize)
    }
}

object CATEGORICAL_INTEGER extends DataType {
    override def isOfType(sampleData: List[String]): Boolean = {
        val categoricalSize: Int = 2
        isCategorical(sampleData, categoricalSize)
    }
}

object EMPTY extends DataType {
    override def isOfType(sampleData: List[String]): Boolean = {
        val emptyRepresentation: Set[String] = Set("N\\A", "\\N", "NULL", "blank", "(blank)", "NaN", "NA")
        matchInDictionary(sampleData, emptyRepresentation.map(_.toLowerCase))
    }
}

